### ISSUE TITLE
Feature: allow PSFs to be saved and reused

### ISSUE THREAD

**Issue by tlambert03** [tlambert03/microsim#23]
PSFs can take a while to generate, but are frequently reused many times over.  So it would be nice to have a way to generate them once, save to disk, and reuse.  This is made slightly complicated by all of the parameters that go into PSF.  That is, if we were to allow the user to pass in a filepath for the PSF, we need to determine whether that PSF actually matches all the corresponding parameters in the simulation and at the very least emit a warning if the provided psf would give invalid results with the rest of the simulation parameters.   

So, a save/reload implementation should have 
1. something like a `Simulation.save_psf()` method (this could conceivably also go on the `Modality` object, but that decision isn't that important, and can be changed later).  This method should save the PSF along with whatever parameters were used.  The format could be `.npy`, or `.tiff`, or `.zarr`, etc... i think we should probably just go with zarr for now, since it's clearer there how we can associate the metadata.
2. A way to declare that the simulation should load a PSF from a file, rather than calculate a new one.  When loading, it should make sure that the params stored in the file match the simulation.

as for *where* the code needs to be added:

- PSFs may be generated in many different places, depending on the modality for example, for confocal it's [here](https://github.com/tlambert03/microsim/blob/main/src/microsim/schema/modality/confocal.py#L28) and widefield does it in the corresponding `Widefield.render` method.
- each of those call `psf.make_psf`, which calls `psf.vectorial_psf_centered()` which calls `psf.vectorial_psf` which calls `vectorial_rz`.  Vectorial rz is a single XZ slice through the center of the PSF, which then gets rotated by `rz_to_xyz` to create the 3D volume.  So, we'd need the code to check whether a psf file is being used, and avoid calling `make_psf` instead.  
    - side note: if it becomes easier to just pass the full `Simulation` object to `Modality.render`, we can change those method signatures.
- technically, all we really need to save to disk is that single rz plane (in which case we'd need to call `rz_to_xyz` each time it gets loaded) but it would also be fine just to save the full/final 3d volume.
- as far as the parameters that get stored and compared on reload it should probably be the arguments to `vectorial_psf`.

cc  @fjug @rayanirban ... hope that helps clarify some of the entry points.  Let me know if you want to take a stab at it, have questions, or want me to do it when I have time :)

**Comment by fjug** [comment]
Quick question: if we wanted to create S samples (maybe 1000) and C confocal setups (maybe 22) and then create S*C (here 22,000) microscopy images. I would now have hoped to be able to do something like this:

1. Create 1000 instances of ms.Sample(...)
2. Create 22 instances of ms.Simulation(...) having the pinhole props etc. we want
3. then something like this
```
for s in S:
    for c in C:
        c.sample = s
        sim.run()
```

If `sim.run()` would only create a PSF if it was not already created (currently not the case if I understand [this here](https://github.com/tlambert03/microsim/blob/main/src/microsim/schema/modality/confocal.py#L28) correctly) then we would already save ourselves a ton of time by never having to re-create PSFs and being able to generate images for many sample in one glorious run.

Please educate me on why my idea is not great (or how we should go about it).

Thanks!

**Comment by tlambert03** [comment]
Your idea is not not great 😊, let's do it... I did a similar thing for caching ground truth in the Simulation object, and we should probably generally be implementing more memoizing of expensive objects.

I can imagine a general memoization scheme here, with a global memory limit, etc... but for now why don't you just do something similar to the quick-and-dirty thing I did in `Simulation.ground_truth`.  This will likely need to be done at the  Modality level, which means it needs to be done in multiple places for now, but it's fine as a stop-gap

**Comment by tlambert03** [comment]
Totally fine.  In parallel ill take a stab at a global mechanism for #25

**Comment by fjug** [comment]
Thanks a lot. We are now back at it. Feedback will come! 😄
